src/main/java/org/folio/services/invoice/PoLinePaymentStatusUpdateService.java

```diff
@@ -109,6 +120,81 @@ public Future<Void> updatePoLinePaymentStatusToCancelInvoice(Invoice invoiceFrom
       });
   }
 
+  /**
+   * Updates payment status of the associated PO Lines (when poLinePaymentStatus is not used)
+   *
+   * @param invoiceLines the invoice lines to be paid
+   * @return CompletableFuture that indicates when transition is completed
+   */
+  private Future<Void> payPoLines(List<InvoiceLine> invoiceLines, RequestContext requestContext) {
+    Map<String, List<InvoiceLine>> poLineIdInvoiceLinesMap = groupInvoiceLinesByPoLineId(invoiceLines);
+    return fetchPoLines(poLineIdInvoiceLinesMap, requestContext)
+      .map(this::updatePoLinesPaymentStatus)
+      .compose(poLines -> orderLineService.updateCompositePoLines(poLines, requestContext));
+  }
+
+  private Map<String, List<InvoiceLine>>  groupInvoiceLinesByPoLineId(List<InvoiceLine> invoiceLines) {
+    if (CollectionUtils.isEmpty(invoiceLines)) {
+      return Collections.emptyMap();
+    }
+    return invoiceLines
+      .stream()
+      .filter(invoiceLine -> StringUtils.isNotEmpty(invoiceLine.getPoLineId()))
+      .collect(Collectors.groupingBy(InvoiceLine::getPoLineId));
+  }
+
+  /**
+   * Retrieves PO Lines associated with invoice lines and calculates expected PO Line's payment status
+   * @param poLineIdsWithInvoiceLines map where key is PO Line id and value is list of associated invoice lines
+   * @return map where key is {@link CompositePoLine} and value is expected PO Line's payment status
+   */
+  private Future<Map<CompositePoLine, CompositePoLine.PaymentStatus>> fetchPoLines(Map<String, List<InvoiceLine>> poLineIdsWithInvoiceLines,
+    RequestContext requestContext) {
+    List<Future<CompositePoLine>> futures = poLineIdsWithInvoiceLines.keySet()
+      .stream()
+      .map(poLine -> orderLineService.getPoLine(poLine, requestContext))
+      .toList();
+
+    return collectResultsOnSuccess(futures)
+      .map(compositePoLines ->  compositePoLines.stream()
+        .collect(Collectors.toMap(poLine -> poLine, poLine -> getPoLinePaymentStatus(poLineIdsWithInvoiceLines.get(poLine.getId())))));
```

Comment:
A bit simplified variant, but can be left as is:
```suggestion
    return collectResultsOnSuccess(futures).map(compositePoLines ->  StreamEx.of(compositePoLines)
      .toMap(Function.identity(), poLine -> getPoLinePaymentStatus(poLineIdsWithInvoiceLines.get(poLine.getId()))));
```

Reply:
This is not new code and I just moved it. I should probably not touch it in this PR, to keep the changes more readable. Additionally, StreamEx is not part of the [officially supported technologies for Trillium](https://folio-org.atlassian.net/wiki/spaces/TC/pages/730891059/Trillium), so I think we should avoid using it.